src/debugger/include/debugger/debugger.h

```diff
@@ -29,47 +30,106 @@ namespace debugger {
 /**
  * Define the debugger class, for breakpoint debugging the CPU and Memory
  */
-class Debugger {
+class DebuggerCore {
 	/**
 	 * Ticks for this object
 	 */
-	unsigned long long ticks;
+	unsigned long long ticks = 0;
 
 	/**
 	 * Gameboy instance
 	 */
 	std::unique_ptr<gameboy::Gameboy> gameboy;
 
 	/**
-	 * Pointer to CPU instance inside gameboy
+	 * A vector container for storing instruction breakpoints
 	 */
-	CPU *cpu;
+	std::vector<Address> breakpoints;
 
 	/**
-	 * Pointer to memory instance inside gameboy
+	 * A bool value to specify whether the debugger has currently encountered a
+	 * breakpoint condition and waiting for the user input. This will reset to
+	 * false once run is called again after a breakpoint.
 	 */
-	Memory *memory;
+	bool is_breaking = false;
 
 	/**
-	 * Pointer to GPU instance inside gameboy
+	 * A vector for storing all the processed breakpoints
 	 */
-	GPU *gpu;
+	std::vector<Address> processed_breakpoints;
 
 	/**
-	 * Pointer to Cartridge instance inside gameboy
+	 * A vector container for storing breakpoints at CPU cycles
 	 */
-	Cartridge *cartridge;
+	std::vector<ClockCycles> cycle_breakpoints;
+
+	/**
+	 * A vector container for storing breakpoints at GameBoy cycles
+	 */
+	std::vector<ClockCycles> tick_breakpoints;
+
+	/**
+	 * A vector container iterator to keep track of breakpoints
+	 */
+	std::vector<Address>::iterator breakpoints_iter = breakpoints.begin();
+
+	/**
+	 * A flag to check if we have reached the breakpoint
+	 */
+	bool is_breakpoint = false;
 
   public:
 	/**
-	 * Debugger constructor
+	 * DebuggerCore constructor
 	 */
-	Debugger(std::unique_ptr<gameboy::Gameboy> gameboy);
+	DebuggerCore(std::unique_ptr<gameboy::Gameboy> gameboy);
 
 	/**
 	 * Run debugger iteration
 	 */
-	void tick();
+	virtual void tick();
+
+	/**
+	 * @brief Adds breakpoints to the Vector Container, skips if breakpoint
+	 * already present
+	 * @param breakpoint of Instructions, ticks and clocks
+	 *
+	 * returns 0 if Breakpoint was already present, so not added;
+	 * returns 1 if Breakpoint was successfully inserted;
+	 * returns 2 default return value. Will not reach here;
+	 * I chose uint8_t as a return type so that we can add to the error codes
```

Comment:
My bad, I'd already added that comment above, but again, why do we need error codes here?

Even if a breakpoint was already present, can't we still just ignore it and return success? At the moment, even a boolean error code doesn't seem necessary.

Reply:
I guess, but I feel it would be better to know what's going on inside. It is kinda unnecessary but it's just doesn't feel right to return success if something's repeated and not added.